pkg/kubelet/dockershim/docker_container.go

```diff
@@ -0,0 +1,195 @@
+/*
+Copyright 2016 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package dockershim
+
+import (
+	"fmt"
+	"io"
+
+	dockertypes "github.com/docker/engine-api/types"
+	dockercontainer "github.com/docker/engine-api/types/container"
+	dockerfilters "github.com/docker/engine-api/types/filters"
+	dockerstrslice "github.com/docker/engine-api/types/strslice"
+	runtimeApi "k8s.io/kubernetes/pkg/kubelet/api/v1alpha1/runtime"
+)
+
+// ListContainers lists all containers matching the filter.
+func (ds *dockerService) ListContainers(filter *runtimeApi.ContainerFilter) ([]*runtimeApi.Container, error) {
+	opts := dockertypes.ContainerListOptions{All: true}
+
+	opts.Filter = dockerfilters.NewArgs()
+	if filter != nil {
+		if filter.Name != nil {
+			opts.Filter.Add("name", filter.GetName())
+		}
+		if filter.Id != nil {
+			opts.Filter.Add("id", filter.GetId())
+		}
+		if filter.State != nil {
+			opts.Filter.Add("status", toDockerContainerStatus(filter.GetState()))
+		}
+		if filter.PodSandboxId != nil {
+			// TODO: implement this after sandbox functions are implemented.
+		}
+
+		if filter.LabelSelector != nil {
+			for k, v := range filter.LabelSelector {
+				opts.Filter.Add(k, v)
+			}
+		}
+	}
+	containers, err := ds.client.ListContainers(opts)
+	if err != nil {
+		return nil, nil
+	}
+	// Convert docker to runtime api containers.
+	result := []*runtimeApi.Container{}
+	for _, c := range containers {
+		result = append(result, toRuntimeAPIContainer(&c))
+	}
+	return result, nil
+}
+
+// CreateContainer creates a new container in the given PodSandbox
+// Note: docker doesn't use LogPath yet.
+// TODO: check if the default values returned by the runtime API are ok.
+func (ds *dockerService) CreateContainer(podSandBoxID string, config *runtimeApi.ContainerConfig, sandboxConfig *runtimeApi.PodSandboxConfig) (string, error) {
+	if config == nil {
+		return "", fmt.Errorf("container config is nil")
+	}
+	if sandboxConfig == nil {
+		return "", fmt.Errorf("sandbox config is nil for container %q", config.GetName())
+	}
+
+	// Merge annotations and labels because docker supports only labels.
+	// TODO: should we apply docker-specific labels?
+	labels := config.GetLabels()
+	for k, v := range config.GetAnnotations() {
+		if _, ok := labels[k]; !ok {
+			// Only write to labels if the key doesn't exist.
+			labels[k] = v
```

Comment:
Use a prefix to distinguish labels and annotations. Or else when get container status how do we know which is label, which is annotation? :)